azkaban-web-server/src/restli/java/azkaban/restli/ProjectManagerResource.java

```diff
@@ -97,22 +101,51 @@ public String deploy(@ActionParam("sessionId") String sessionId,
       // complete.
       logger.info("Downloading package from " + packageUrl);
       FileUtils.copyURLToFile(url, archiveFile);
-      Props props = new Props();
 
       logger.info("Downloaded to " + archiveFile.toString());
-      projectManager.uploadProject(project, archiveFile, "zip", user, props);
     } catch (IOException e) {
       String errorMsg =
           "Download of URL " + packageUrl + " to " + archiveFile.toString()
               + " failed";
       logger.error(errorMsg, e);
-      throw new ProjectManagerException(errorMsg, e);
+      if (tempDir.exists()) {
+        FileUtils.deleteDirectory(tempDir);
+      }
+      throw new RestLiServiceException(HttpStatus.S_400_BAD_REQUEST, errorMsg, e);
+    }
+
+    try {
+      // Check if project upload runs into any errors, such as the file
+      // having blacklisted jars
+      Props props = new Props();
+      Map<String, ValidationReport> reports = projectManager.uploadProject(project, archiveFile, "zip", user, props);
+      checkReport(reports);
+      return Integer.toString(project.getVersion());
+    } catch (ProjectManagerException e) {
+      String errorMsg = "Upload of project " + project + " from " + archiveFile + " failed";
+      logger.error(errorMsg, e);
+      throw e;
     } finally {
       if (tempDir.exists()) {
         FileUtils.deleteDirectory(tempDir);
       }
     }
-    return Integer.toString(project.getVersion());
+  }
+
+  void checkReport(Map<String, ValidationReport> reports) throws RestLiServiceException {
```

Comment:
nit: checkReports seems more apt. You are taking a collection as an argument as opposed to a single Report object